lib/validator/validator.go

```diff
@@ -288,31 +293,32 @@ func validateHostEndpointSpec(v *validator.Validate, structLevel *validator.Stru
 	}
 }
 
-func validatePoolMetadata(v *validator.Validate, structLevel *validator.StructLevel) {
-	pool := structLevel.CurrentStruct.Interface().(api.IPPoolMetadata)
-
-	// The Calico IPAM places restrictions on the minimum IP pool size, check that the
-	// pool is at least the minimum size.
-	if pool.CIDR.IP != nil {
-		ones, bits := pool.CIDR.Mask.Size()
-		log.Debugf("Pool CIDR: %s, num bits: %d", pool.CIDR, bits-ones)
-		if bits-ones < 6 {
-			structLevel.ReportError(reflect.ValueOf(pool.CIDR),
-				"CIDR", "", reason("IP pool is too small"))
+func validateIPPool(v *validator.Validate, structLevel *validator.StructLevel) {
+	pool := structLevel.CurrentStruct.Interface().(api.IPPool)
+
+	// Validation of the data occurs before checking whether Metadata
+	// fields are complete, so need to check whether CIDR is assigned before
+	// performing cross-checks.  If CIDR is not assigned this will be
+	// picked up during Metadata->Key conversion.
+	if pool.Metadata.CIDR.IP != nil {
+		// IPIP cannot be enabled for IPv6.
+		if pool.Metadata.CIDR.Version() == 6 && pool.Spec.IPIP != nil && pool.Spec.IPIP.Enabled {
+			structLevel.ReportError(reflect.ValueOf(pool.Spec.IPIP.Enabled),
+				"IPIP.Enabled", "", reason("IPIP is not supported on an IPv6 IP pool"))
 		}
-	}
-}
-
-func validateIPIPConfiguration(v *validator.Validate, structLevel *validator.StructLevel) {
-	ipip := structLevel.CurrentStruct.Interface().(api.IPIPConfiguration)
 
-	log.Debugf("Validate IPIP: Enabled %b, Mode: %s", ipip.Enabled, ipip.Mode)
-	if ipip.Enabled {
-		if !ipipModeRegex.MatchString(string(ipip.Mode)) {
-			structLevel.ReportError(reflect.ValueOf(ipip.Mode),
-				"Mode", "", reason("When IPIP Enabled, Mode should be always, <empty string>, or cross-subnet"))
+		// The Calico IPAM places restrictions on the minimum IP pool size.  If
+		// the pool is enabled, check that the pool is at least the minimum size.
+		if !pool.Spec.Disabled {
+			ones, bits := pool.Metadata.CIDR.Mask.Size()
+			log.Debugf("Pool CIDR: %s, num bits: %d", pool.Metadata.CIDR, bits-ones)
+			if bits-ones < 6 {
+				structLevel.ReportError(reflect.ValueOf(pool.Metadata.CIDR),
+					"CIDR", "", reason("IP pool CIDR is too small to an IPAM enabled pool"))
```

Comment:
I'm confused by this error message wording. What do you mean by `an IPAM enabled pool`? Also, we should mention the minimum pool size in the message so user can know what is the minimum size if they see this error.